clang/test/CodeGenSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,163 @@
+// CHECK: call spir_func void @_Z18kernel_single_taskIZ4mainE7kernel3Z4mainEUlPZ4mainEUlvE0_E_EvT0_
+//
+extern "C" void printf(const char *) {}
+
+template <typename T>
+void template_param() {
+  printf(__builtin_unique_stable_name(T));
+}
+
+template <typename T>
+T getT() { return T{}; }
+
+template <typename T>
+void lambda_in_dependent_function() {
+  auto y = [] {};
+  printf(__builtin_unique_stable_name(y));
+}
+
+template <typename Tw, typename Tz>
+void lambda_two_dep() {
+  auto z = [] {};
+  printf(__builtin_unique_stable_name(z));
+}
+
+template <typename Tw, typename Tz>
+void lambda_no_dep(Tw a, Tz b) {
+  auto p = [](Tw a, Tz b) { return ((Tz)a + b); };
+  printf(__builtin_unique_stable_name(p));
+}
+
+#define DEF_IN_MACRO()                           \
+  auto MACRO_X = []() {};                        \
+  auto MACRO_Y = []() {};                        \
+  printf(__builtin_unique_stable_name(MACRO_X)); \
+  printf(__builtin_unique_stable_name(MACRO_Y));
+
+#define MACRO_CALLS_MACRO() \
+  { DEF_IN_MACRO(); }       \
+  { DEF_IN_MACRO(); }
+
+template <typename Ty>
+auto func() -> decltype(__builtin_unique_stable_name(Ty::str));
+// CHECK: declare spir_func i8* @_Z4funcI4DerpEDTu20__unique_stable_nameXsrT_3strEEEv()
+
+struct Derp {
+  static constexpr const char str[] = "derp derp derp";
+};
+
+#define IF_MACRO()              \
+  auto l1 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l1);
+
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+  kernel_single_task<class kernel>(
+      []() {
+        printf(__builtin_unique_stable_name(int));
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[INT_SIZE]], [[INT_SIZE]]* @[[INT]]
+
+        auto x = []() {};
+        printf(__builtin_unique_stable_name(x));
+        printf(__builtin_unique_stable_name(decltype(x)));
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[LAMBDA_X_SIZE]], [[LAMBDA_X_SIZE]]* @[[LAMBDA_X]]
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[LAMBDA_Y_SIZE]], [[LAMBDA_Y_SIZE]]* @[[LAMBDA_Y]]
+
+        DEF_IN_MACRO();
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[MACRO_SIZE]], [[MACRO_SIZE]]* @[[MACRO_X]]
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[MACRO_SIZE]], [[MACRO_SIZE]]* @[[MACRO_Y]]
+
+        MACRO_CALLS_MACRO();
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[MACRO_MACRO_SIZE]], [[MACRO_MACRO_SIZE]]* @[[MACRO_MACRO_X]]
+        // CHECK: call spir_func void @printf(i8* getelementptr inbounds ([[MACRO_MACRO_SIZE]], [[MACRO_MACRO_SIZE]]* @[[MACRO_MACRO_Y]]
+
+        template_param<int>();
+        // CHECK: define linkonce_odr spir_func void @_Z14template_paramIiEvv
```

Comment:
Instead of this pattern here, do a CHECK on the call to the templ_param function, then move these two lines below this function.  Otherwise this is going to be REALLY fragile, since these 'check' lines move FileCheck outside this function, so you won't be able to test anything after this inside this lambda.

Reply:
Erich, I am not sure I follow this suggestion correctly. Are you suggesting adding the call to the templ_param function at the beginning of the file? 

```  
     template_param<int>();
        // CHECK: call spir_func void @_Z14template_paramIiEvv
        // CHECK: define linkonce_odr spir_func void @_Z14template_paramIiEvv
        // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[INT1_SIZE]], [[INT1_SIZE]]* @[[INT1]]
```
this works for this case but not the next one: 

```
        template_param<decltype(x)>();
        // : call spir_func void @_Z14template_paramIZZ4mainENKUlvE_clEvEUlvE_Evv  // doesn't match
        // CHECK: define internal spir_func void @_Z14template_paramIZZ4mainENKUlvE_clEvEUlvE_Evv
        // CHECK: call spir_func void @puts(i8* getelementptr inbounds ([[LAMBDA_SIZE]], [[LAMBDA_SIZE]]* @[[LAMBDA]]
```

I understand the part about fragility. Struggling with it again, now that I moved the test case just a little bit. :(

Reply:
Almost.  Think about how the IR is going to be laid out here.  ALL of the lambda is going to be there in 1 place, followed by the bodies of the other functions.  Additionally, FileCheck works in line-order, so once you match a line, you can't really match lines before it. 

So if you do those 2nd 2 CHECK lines in template_param<int>();, the FileCheck "Cursor" is now way at the end.  Instead, you should do ONLY the 1st one inside the lambda-body, and 'CHECK' the contents of the function later, after the file's body.

Reply:
Makes sense. Thank you.

Reply:
Could you please take a look at my new patch? It might not be exactly what you suggested, but I grouped all the CHECKs for spir_func "template_param" calls and moved them above, leaving  the definition and @puts calls behind.

Reply:
No,still doesnt look right.  ONLY the all to the function template should be here, the 'define' and the 'call printf' lines need to be at the end, down near where they happen in the actual llvm-IR.  I'd suggest running the IR line only (sans-filecheck) to see how it is laid out, and you'll see how they should be.

Reply:
Right. That sounds better. Then the CHECK line sits next to the actual call to the function template. Thanks!